Add scripts to analyse a sites query param usage

dsingleton opened the pull request:
By default transition mappings will ignore any query parameters, as
they most commonly don't affect the semantic meaning of the URL.

There may be some exceptions, eg, `document.php?id=foo` that we need
to explicitly configure a transition site to ignore.

These scripts are used to work out, from a list of URLs for s site,
which query params may be be meanginful, based on frequency of use
and which URLs rely most on query params.

Note: Some of the existing scripts take a file as an argument, i've
expanded this support stdin/pipes, as i've found it really useful
to tweak input (filtering out terms, etc) without having to create
a real file each time.

dsingleton commented:
(@mikejustdoit wrote most of `analyse_query_params.sh`, i've tweaked some POSIX bits, and documented it)

jennyd commented on tools/analyse_query_params.sh:
```diff
@@ -0,0 +1,34 @@
+#
+#
+# For the input, a list of URLs, shows which are most common query
+# parameters, Eg, those that may need to be exempt from normalisation
```
Instead of `exempt from normalisation` here, could we say `preserved by canonicalization` to keep to more standard transition terms?

jennyd commented on tools/analyse_query_params.sh:
```diff
@@ -0,0 +1,34 @@
+  sed 's/&amp;/\&/g' | \
+  # Split query params into one pair (`foo=bar`) per line
+  tr '&' '\n' | \
+  # Take the value of the pair (`bar` for `foo=bar`)
```
I find this comment a bit confusing - we're keeping the key, not the value, but 'Take' suggests the opposite to me.

jennyd commented on tools/analyse_query_params.sh:
```diff
@@ -0,0 +1,34 @@
+  # Take the value of the pair (`bar` for `foo=bar`)
+  cut -d= -f1 | \
+  # count the frequency of each value, highest first
+  sort | uniq -c | sort -rn
```
8 pipes :8ball: It's great having comments for every line for this kind of thing.

jennyd commented on tools/analyse_query_usage.sh:
```diff
@@ -0,0 +1,25 @@
+#
+# For the input, a list of URLs, shows which are most common when you
+# ignore all query parameters. Eg, those that will be normalised in
+# mappings unless some query parameters are marked as important.
```
For the same reason as above, could the second sentence here be something like 'the number of URLs that will be reduced to that single URL by canonicalization unless some query parameters are marked as significant'?

jennyd commented on tools/analyse_query_params.sh:
```diff
@@ -0,0 +1,34 @@
+  sed 's/&amp;/\&/g' | \
+  # Split query params into one pair (`foo=bar`) per line
+  tr '&' '\n' | \
+  # Take the key of the pair (`bar` for `foo=bar`)
```
We're keeping `foo` here, not `bar`... :football: :bar_chart:

jennyd commented on tools/analyse_query_usage.sh:
```diff
@@ -0,0 +1,25 @@
+# Input is either file arg contents or stdin
+cat "$@" | \
+  # Take all of the URL leading up to `?`, or all if not present
+  cut -d? -f1 | \
```
The output at this stage can include URLs which are identical apart from with/without trailing slashes, which would be removed by canonicalization. Could we drop trailing slashes to further condense the output to be more similar to canonicalization?

dsingleton replied:
I thought about that, and tbh it's one of the reasons I added support fir pipable input, as I was doing it separately.

I _think_ it would be better for this script to only be concerned with normalising by query params, and if URL normalisation (which could potentially include other things) is required, then to do that to the input before this script gets it. Feels more unix-y?

jennyd replied:
Doing that on the input to this script would mean stripping out trailing slashes from the middle of the URL (because the input contains query strings) rather than dropping them from the ends of lines here. You're right, though - there are other changes made by canonicalization which this script shouldn't need to make (downcasing, encoding etc) so it makes sense to not do any of them here.

dsingleton commented:
Fixed typos and added the improved doc. 

Thank you for the feedback on that, it makes a lot more sense now :+1: 

Commented on the trailing `/` issue, but i'm not sure it should be a concern of this script. If we can resolve that then this is mergable, I think?
